V4 support

liquid-8 opened the pull request:
Basic functionality, like price fetching, quoting, and making simple swaps etc

liquid-8 committed:
v4 support pre-alpha

liquid-8 committed:
v4 support pre-alpha

liquid-8 committed:
Revert "v4 support pre-alpha"

This reverts commit fa7563256d3c4955035dc6727ad1e185d1e978c6.

liquid-8 committed:
Revert "v4 support pre-alpha"

This reverts commit de63ce45856bec5556d62f9f832df4191148c162.

liquid-8 committed:
Merge branch 'uniswap-python:master' into master

liquid-8 committed:
Merge branch 'uniswap-python:master' into master

liquid-8 committed:
Update __init__.py

liquid-8 committed:
Merge pull request #417 from liquid-8/master

fin

liquid-8 committed:
Add files via upload

liquid-8 committed:
Add Uniswap4 to module exports

liquid-8 committed:
Delete uniswap/uniswap-v4 directory

liquid-8 committed:
Add files via upload

liquid-8 commented on uniswap/uniswap4.py:
```diff
@@ -0,0 +1,457 @@
+        self.router = _load_contract(self.w3, abi_name="uniswap-v4/router", address=self.router_address)
+        self.stateview = _load_contract(self.w3, abi_name="uniswap-v4/stateview", address=self.stateview_address)
+        self.permit2 = _load_contract(self.w3, abi_name="uniswap-v4/permit2", address=self.permit2_address)
+        weth_by_dex[dex_name] = WRAPPED_ETH_ADDRESS
```
need to refactor this and related

liquid-8 committed:
Refactor uniswap4.py by removing unused imports

Removed unused imports and cleaned up code formatting.

liquid-8 committed:
Add files via upload

Pool manager contract ABI

liquid-8 committed:
Delete uniswap/configs/evmuniV4_permit2.ini

Deleted due to naming fix

liquid-8 committed:
Delete uniswap/configs/evmuniV4_chains.ini

Deleted due to naming fix

liquid-8 committed:
Delete uniswap/configs/evmuniV4_quoter.ini

Deleted due to naming fix

liquid-8 committed:
Delete uniswap/configs/evmuniV4_router.ini

Deleted due to naming fix

liquid-8 committed:
Delete uniswap/configs/evmuniV4_stateview.ini

Deleted due to naming fix

liquid-8 committed:
Add files via upload

Naming fix

liquid-8 committed:
Add files via upload

Liquidity pools list management class added, UniswapV4 improvements.

ErikBjare reviewed (approved):
Generally approved. Haven't checked why CI is failing yet.

Could use some auto-formatting by ruff.

ErikBjare commented:
Ah, I see CI is failing due to outdated actions. I think we can merge this and address the outdated CI and any CI issues related to this PR in follow-up.

ErikBjare commented on uniswap/configs/chains.ini:
Do we really want `ini`s for this? Are we expecting users to be able to modify them? Why not just use a global dict in a `.py`?

liquid-8 replied:
I'm going to add option to configure this to any folder, then users can add any networks themselves not touching code

liquid-8 replied:
I'm going to add option to configure this to any folder, then users can add any networks themselves not touching code

ErikBjare commented on uniswap/uniswap4.py:
```diff
@@ -0,0 +1,576 @@
+        Generic parameters builder for universal router execute() call.
+        WIP
+        """
+        pass
```
Unused

liquid-8 committed:
Delete uniswap/uni4base.py

refactor

liquid-8 committed:
Delete uniswap/configs/chains.ini

refactor

liquid-8 committed:
Delete uniswap/configs directory

refactor

liquid-8 committed:
Beta

refactoring+bug fixes

ErikBjare commented:
@greptileai review

ErikBjare commented on uniswap/uniswap4.py:
Maybe move the v4 stuff into e.g. `uniswap/v4/uniswap.py` for better structure

liquid-8 committed:
Add files via upload

liquid-8 committed:
Add files via upload

liquid-8 committed:
Delete uniswap/v4pools.py

moved to util

liquid-8 committed:
Delete uniswap/v4types.py

moved to types

liquid-8 committed:
Delete uniswap/v4constants.py

moved to constants

liquid-8 committed:
Add files via upload

Structure refactoring; fixes; StateView contract support added

ErikBjare commented on uniswap/uniswap4.py:
```diff
@@ -0,0 +1,933 @@
+        max_slippage: float = 0.1,
+        max_gas: float = 250000.0,
+        max_gprice: float = 1.80,
+        london_fork: int = 1,
```
Make bool?

ErikBjare commented on uniswap/uniswap4.py:
```diff
@@ -0,0 +1,933 @@
+        version: int = 4,
+        max_slippage: float = 0.1,
+        max_gas: float = 250000.0,
+        max_gprice: float = 1.80,
```
```suggestion
        max_gas_price: float = 1.80,
```

ErikBjare commented on uniswap/uniswap4.py:
```diff
@@ -0,0 +1,933 @@
+        self.max_approval_check_hex = f"0x{15 * '0'}{49 * 'f'}"
+        self.max_approval_check_int = int(self.max_approval_check_hex, 16)
+        self.gas_limit = max_gas
+        self.gas_price = max_gprice
```
Actually, maybe just make it all consistently "gas_limit" and "gas_price"

ErikBjare commented on uniswap/uniswap4.py:
```diff
@@ -0,0 +1,933 @@
+        self.gas_limit = max_gas
+        self.gas_price = max_gprice
+        self.london_style = london_fork
+        self.london_priorityfee = max_priorityfee
```
self.priority_fee, no need to mention london

ErikBjare commented on uniswap/util.py:
Strange that this diff became so messy

ErikBjare commented on uniswap/types.py:
```diff
@@ -1,5 +1,13 @@
 AddressLike = Union[Address, ChecksumAddress]
+
+@dataclass
+class pool_key:
```
```suggestion
class PoolKey:
```

ErikBjare commented on uniswap/types.py:
```diff
@@ -1,5 +1,13 @@
+    currency1 : str
+    fee : int
+    tick_spacing : int
+    hooks : str
```
Not auto-formatted? I guess we don't have pre-commit hooks set up in this repo? (we should probably adopt ruff and ruff-format)

liquid-8 replied:
That was black default auto formatting, will try ruff. Surprised with that mess too

ErikBjare commented on uniswap/constants.py:
Strange that this diff too became super messy, maybe you're changing the line endings from \n to \n\r? (Windows?)

pre-commit with ruff-format would fix this I think

liquid-8 replied:
Just the corrupted local settings, fixed that and diff mess. Need to set up linux VM def, this becomes annoying

ErikBjare reviewed (changes requested):
Nice, better structure!

I think it would be worthwhile to add a proper pre-commit/prek config with ruff and ruff-format though, would probably fix the weird diffs that suggest a lot more has changed than it has.

liquid-8 committed:
Formmating fixes, naming fixes

liquid-8 committed:
fix constants

liquid-8 committed:
Fix util

liquid-8 committed:
technical commit

liquid-8 committed:
Technical fixes

liquid-8 committed:
Merge branch 'dev/v4-fin' into uniswap4_RC

liquid-8 committed:
Merge pull request #419 from liquid-8/uniswap4_RC

technical commit/fixes

liquid-8 committed:
pre-commit formatter/linter

liquid-8 committed:
Merge branch 'uniswap4_RC' of https://github.com/liquid-8/uniswap-python into uniswap4_RC

liquid-8 committed:
fix pyproject.toml

liquid-8 committed:
Merge pull request #420 from liquid-8/uniswap4_RC

pre-commit linter/formatter

liquid-8 commented:
> Nice, better structure!
> 
> I think it would be worthwhile to add a proper pre-commit/prek config with ruff and ruff-format though, would probably fix the weird diffs that suggest a lot more has changed than it has.

done

ErikBjare committed:
ci: update GitHub Actions to supported versions

ErikBjare committed:
ci: fix python version output quoting in workflow

ErikBjare committed:
build(deps): update poetry.lock

ErikBjare commented:
Lots of failing typechecks.

I fixed the basics that were breaking in the CI runs, but the tests are still failing due to the API key for the node seemingly being heavily rate limited. I'll look into an alternate node provider.

liquid-8 committed:
typecheck fixes

liquid-8 committed:
review fixes

liquid-8 committed:
Merge pull request #421 from liquid-8/uniswap4_RC

typecheck fixes; web3 version raised to 6.5.0 for contract.Event().get_logs() support

liquid-8 committed:
poetry.lock update

liquid-8 committed:
Merge branch 'dev/v4-fin' into uniswap4_RC

liquid-8 committed:
Merge pull request #422 from liquid-8/uniswap4_RC

poetry.lock update

liquid-8 committed:
more typecheck fixes

liquid-8 committed:
Merge branch 'uniswap4_RC' of https://github.com/liquid-8/uniswap-python into uniswap4_RC

liquid-8 committed:
Merge pull request #423 from liquid-8/uniswap4_RC

more typecheck fix

liquid-8 committed:
more typecheck

liquid-8 committed:
get_token fix

liquid-8 committed:
Merge pull request #424 from liquid-8/uniswap4_RC

more typecheck

liquid-8 commented:
typecheck fixed

liquid-8 committed:
PositionDescriptor support; PositionManager read methods support; misc

liquid-8 committed:
fixes

liquid-8 committed:
Merge pull request #425 from liquid-8/uniswap4_RC

PositionDescriptor support; PositionManager read methods support; misc

liquid-8 committed:
PositionManager full support; PoolManager full support; contracts-related methods naming improvement.

liquid-8 committed:
fixes

liquid-8 committed:
Merge pull request #426 from liquid-8/uniswap4_RC

PositionManager full support; PoolManager full support; contracts-related methods naming improvements.

ErikBjare commented:
@greptileai review

ErikBjare commented on uniswap/uniswap4.py:
```diff
@@ -0,0 +1,2096 @@
+        provider: Optional[str] = None,
+        web3: Optional[Web3] = None,
+        version: int = 4,
+        max_slippage: float = 0.1,
```
Should be 0.01 or at least 0.05, and might be a good thing to test for.

ErikBjare commented on uniswap/uniswap4.py:
```diff
@@ -0,0 +1,2096 @@
+        :param provider: Can be optionally set to a Web3 provider URI. If none set, will fall back to the PROVIDER environment variable, or web3 if set.
+        :param web3: Can be optionally set to a custom Web3 instance.
+        :param version: Which version of the Uniswap contracts to use.
+        :param max_slippage: Maximum slippage for a trade, as a float (0.01 is 1%). WARNING: slippage is untested.
```
Could be tested

ErikBjare commented on uniswap/uniswap4.py:
```diff
@@ -0,0 +1,2096 @@
+        if self.version == 4:
+            tick_info: Dict = self.stateview.functions.getTickInfo(pool_id, tick).call()
+        else:
+            raise ValueError("Function is not supported for this version")
```
Since this class is V4-only, can't we remove all these checks? There are a lot of them...

ErikBjare commented on uniswap/uniswap4.py:
```diff
@@ -0,0 +1,2096 @@
+            _str_to_addr(address) if isinstance(address, str) else address
+        )
+        self.private_key = private_key
+        self.version = version
```
Remove version arg and member?

ErikBjare commented on uniswap/uniswap4.py:
```diff
@@ -0,0 +1,2096 @@
+        private_key: Optional[str],
+        provider: Optional[str] = None,
+        web3: Optional[Web3] = None,
+        version: int = 4,
```
Remove?

ErikBjare commented on uniswap/uniswap4.py:
```diff
@@ -0,0 +1,2096 @@
+        # Give an exchange/router max approval of a token.
+        max_approval = 2**100 - 1
+        expiration: int = int(10**12)
+        print(f"Setting permit for {_addr_to_str(token)} at router contract...")
```
I think most/all of these should probably use:

```
logger = getLogger()
...
logger.info(msg)
```

That way consumers of the library can manage the logging/output of the module better.

ErikBjare commented on pyproject.toml:
```diff
@@ -24,10 +24,11 @@ unipy = "uniswap:main"
 click = "^8.0.3"
 python-dotenv = "*"
 typing-extensions = "*"
+pre-commit = "*"
```
Could be a dev-dependency, but shouldn't be a main dependency. Although I've recently preferred `prek`

ErikBjare reviewed (commented):
Looks good, had a few comments that would clean it up.

I guess now is a good time to write some basic tests for it. I should look into fixing the test CI...

liquid-8 commented:
> Looks good, had a few comments that would clean it up.
> 
> I guess now is a good time to write some basic tests for it. I should look into fixing the test CI...

Suggestions are fine. I was going to merge this class with the main one eventually, but even now, it is just too big. Thanks for the review, will fix and proceed.

liquid-8 committed:
review fixes

liquid-8 committed:
extra fix

liquid-8 committed:
Merge pull request #427 from liquid-8/uniswap4_RC

review fixes

liquid-8 commented:
> > Looks good, had a few comments that would clean it up.
> > I guess now is a good time to write some basic tests for it. I should look into fixing the test CI...
> 
> Suggestions are fine. I was going to merge this class with the main one eventually, but even now, it is just too big. Thanks for the review, will fix and proceed.

And it's done.

liquid-8 committed:
poetry.lock fix

liquid-8 committed:
Merge pull request #428 from liquid-8/uniswap4_RC

poetry.lock fix

ErikBjare commented on uniswap/constants.py:
```diff
@@ -13,6 +13,9 @@
 ETH_ADDRESS = "0x0000000000000000000000000000000000000000"
 WETH9_ADDRESS = "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2"
+ZERO_HOOK = "0x0000000000000000000000000000000000000000"
+WRAPPED_ETH_ADDRESS = "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2"
```
Same as WETH9 above?

liquid-8 replied:
The former is added for readability and clarity, so let it be; agreed regarding the latter tho

liquid-8 committed:
v4 test prototype

liquid-8 committed:
Update uniswap/uniswap4.py

Co-authored-by: greptile-apps[bot] <165735046+greptile-apps[bot]@users.noreply.github.com>

liquid-8 committed:
Update tests/v4tests.py

Co-authored-by: greptile-apps[bot] <165735046+greptile-apps[bot]@users.noreply.github.com>

liquid-8 committed:
fix

liquid-8 committed:
Merge branch 'uniswap4_RC' of https://github.com/liquid-8/uniswap-python into uniswap4_RC

liquid-8 committed:
Merge pull request #429 from liquid-8/uniswap4_RC

v4 test prototype

liquid-8 committed:
tech commit

liquid-8 committed:
tech fix

TimeToBuildBob committed:
fix(ci): replace rate-limited Infura with public RPC (#431)

* fix(ci): replace rate-limited Infura with public RPC and fix type errors

- Switch PROVIDER_MAINNET fallback from Infura to ethereum.publicnode.com
  (Infura public key was hitting "Too Many Requests" rate limits)
- Add dev/v4 and dev/v4-fin branches to CI triggers
- Fix mypy errors: use str() for bytes fallback in token name/symbol
  decode, add type: ignore for web3 get_logs attribute

Co-authored-by: Bob <bob@superuserlabs.org>

* fix(ci): switch Arbitrum RPC and resolve mypy warning

TimeToBuildBob committed:
fix(ci): stop push runs from preferring stale mainnet secret (#432)

liquid-8 committed:
Added multi-hop quotes support

liquid-8 committed:
review fixes

liquid-8 committed:
Merge pull request #433 from liquid-8/uniswap4_RC

Added multi-hop quotes support

liquid-8 committed:
Checks improvements, small logic fix, multi-hop tests added

liquid-8 committed:
Merge pull request #434 from liquid-8/uniswap4_RC

Checks improvements, small logic fix, multi-hop tests added

liquid-8 committed:
Added multi-hop swaps support; added market pricing functions similar to main class; removed hardcoded zero values regarding hooks; impoved pool data fetching process;  misc improvement

liquid-8 committed:
review fix

liquid-8 committed:
Merge branch 'dev/v4-fin' into uniswap4_RC

liquid-8 committed:
Merge pull request #435 from liquid-8/uniswap4_RC

Added multi-hop swaps support etc

liquid-8 committed:
typecheck fix

liquid-8 committed:
Merge branch 'uniswap4_RC' of https://github.com/liquid-8/uniswap-python into uniswap4_RC

liquid-8 committed:
Improved exceptions handling; get_logs() glitch fix

liquid-8 committed:
Merge pull request #436 from liquid-8/uniswap4_RC

typecheck fix

liquid-8 committed:
Added swap tests; added liquidity read functions; added liquidity tests

liquid-8 committed:
typo fix

liquid-8 committed:
Merge pull request #437 from liquid-8/uniswap4_RC

Added swap tests; added liquidity read functions; added liquidity tests

liquid-8 commented:
@greptileai review

liquid-8 committed:
Fixes: get_poolkeys_sublist() - addresses comparison; estimate_price_impact() - route param removed;  decode_position_info() - imcorrect input value handling

liquid-8 committed:
Fix: decode_position_info() - truncated poolID decoding

liquid-8 committed:
Merge pull request #438 from liquid-8/uniswap4_RC

Review fixes

liquid-8 committed:
Add get_position_value() and get_minted_token_id(); related tests; tick math improved; misc

liquid-8 committed:
get_amounts_for_liquidity_by_ticks() refactoring; review suggestions added

liquid-8 committed:
typecheck glitch fix

liquid-8 committed:
typo

liquid-8 committed:
typo

liquid-8 committed:
Merge pull request #439 from liquid-8/uniswap4_RC

Add get_position_value() and get_minted_token_id(); related tests; tick math improved; misc

liquid-8 committed:
get_sqrt_ratio_at_tick() moved to util.py as it is shared with v3;  get_minted_token_id() refactoring to support multiple mints

liquid-8 committed:
review suggestion

liquid-8 committed:
review suggestion

liquid-8 committed:
Merge pull request #440 from liquid-8/uniswap4_RC

get_sqrt_ratio_at_tick() moved to util.py as it is shared with v3;  get_minted_token_id() refactoring to support multiple mints

liquid-8 committed:
Add liquidity management methods; add router commands and pool actions constants

liquid-8 committed:
quick self-fix

liquid-8 committed:
typos

liquid-8 committed:
typos

liquid-8 committed:
Merge pull request #441 from liquid-8/uniswap4_RC

Add liquidity management methods; add router commands and pool actions constants

liquid-8 committed:
Changed direct contract call methods naming to improve usability

liquid-8 committed:
typo

liquid-8 committed:
Merge pull request #442 from liquid-8/uniswap4_RC

Changed direct contract call methods naming to improve usability

liquid-8 commented:
@greptileai review

liquid-8 committed:
Improve log processing in fetch_poolkey_data(); typos fixed

liquid-8 committed:
Merge pull request #443 from liquid-8/uniswap4_RC

Improve log processing in fetch_poolkey_data(); typos fixed

liquid-8 commented:
@ErikBjare take a look and check discord, pls

liquid-8 committed:
Add fetch_poolkey_data()  reconnect functionality

liquid-8 committed:
Merge pull request #444 from liquid-8/uniswap4_RC

Add fetch_poolkey_data()  reconnect functionality

liquid-8 committed:
Change fetch_poolkey_data(): param/var renamed and docstring changed for better clarity; tests exceptions handling improved slightly.

liquid-8 committed:
Merge pull request #445 from liquid-8/uniswap4_RC

Change fetch_poolkey_data(): param/var renamed and docstring changed for better clarity; tests exceptions handling improved slightly.

liquid-8 committed:
Add universal router prerequisites

liquid-8 committed:
quick fix

liquid-8 committed:
Merge pull request #446 from liquid-8/uniswap4_RC

Add universal router prerequisites

ErikBjare commented on uniswap/uniswap4.py:
```diff
@@ -0,0 +1,2874 @@
+        """
+        Helper function to calculate the amount of liquidity that can be provided for a given amount of `token1` and price range defined by `sqrt_ratio_a_x96` and `sqrt_ratio_b_x96`.
+        """
+        if sqrt_ratio_a_x96 > sqrt_ratio_b_x96:
+            sqrt_ratio_a_x96, sqrt_ratio_b_x96 = sqrt_ratio_b_x96, sqrt_ratio_a_x96
+
+        liquidity: int = (amount1 * Q96) // (sqrt_ratio_b_x96 - sqrt_ratio_a_x96)
+        return liquidity
+
+    def get_liquidity_for_amounts(
+        self,
+        sqrt_ratio_a_x96: int,
+        sqrt_ratio_b_x96: int,
+        sqrt_ratio_current_x96: int,
+        amount0: int,
+        amount1: int,
+    ) -> int:
+        """
+        Helper function to calculate the amount of liquidity that can be provided for given amounts of `token0` and `token1` and price range defined by `sqrt_ratio_a_x96` and `sqrt_ratio_b_x96`.
+        """
+        if sqrt_ratio_a_x96 > sqrt_ratio_b_x96:
+            sqrt_ratio_a_x96, sqrt_ratio_b_x96 = sqrt_ratio_b_x96, sqrt_ratio_a_x96
+
+        if sqrt_ratio_current_x96 <= sqrt_ratio_a_x96:
+            liquidity: int = self.get_liquidity_for_amount0(
+                sqrt_ratio_a_x96, sqrt_ratio_b_x96, amount0
+            )
+        elif sqrt_ratio_current_x96 < sqrt_ratio_b_x96:
+            liquidity0: int = self.get_liquidity_for_amount0(
+                sqrt_ratio_current_x96, sqrt_ratio_b_x96, amount0
+            )
+            liquidity1: int = self.get_liquidity_for_amount1(
+                sqrt_ratio_a_x96, sqrt_ratio_current_x96, amount1
+            )
+            liquidity = min(liquidity0, liquidity1)
+        else:
+            liquidity = self.get_liquidity_for_amount1(
+                sqrt_ratio_a_x96, sqrt_ratio_b_x96, amount1
+            )
+        return liquidity
```
Many of these don't need `self` and should be pure functions or `@staticmethod`

ErikBjare reviewed (commented):
LGTM, but Greptile probably disagrees 🙃 

Needs a v4 job in CI with basic tests.

liquid-8 committed:
Add universal router support

liquid-8 committed:
Review fixes

liquid-8 committed:
Merge pull request #447 from liquid-8/uniswap4_RC

Add universal router support

liquid-8 committed:
review suggestions

liquid-8 committed:
Review suggestions

liquid-8 committed:
Merge pull request #448 from liquid-8/uniswap4_RC

Review suggestions

liquid-8 commented:
@greptileai review

liquid-8 committed:
Review suggestions

liquid-8 committed:
Merge pull request #449 from liquid-8/uniswap4_RC

Review suggestions

liquid-8 commented:
@greptileai review

liquid-8 committed:
Fixed custom recipient' logic in swaps; fixed ABIs constants.

liquid-8 committed:
review fixes

liquid-8 committed:
Merge pull request #450 from liquid-8/uniswap4_RC

Fixed custom recipient' logic in swaps; fixed ABIs constants.

liquid-8 commented:
@greptileai review

liquid-8 committed:
Removed custom recipent in swaps; tests added.

liquid-8 committed:
quick fix

liquid-8 committed:
Merge pull request #451 from liquid-8/uniswap4_RC

Removed custom recipent in swaps; tests added.

liquid-8 commented:
@greptileai review

liquid-8 committed:
misc fixes

liquid-8 committed:
Merge pull request #452 from liquid-8/uniswap4_RC

misc fixes

liquid-8 commented:
@ErikBjare Something went wrong w/tests, could you check Discord please?

liquid-8 committed:
Switch from ganachr to foundry-anvil for v4 tests

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
Merge pull request #453 from liquid-8/uniswap4_RC

Switch from ganache to foundry-anvil for v4 tests; tests' structure improved.

liquid-8 committed:
Add custom transactions' nonce; custom delay time within approve(); universal router' commands and actions ABIs clarified

liquid-8 committed:
fix

liquid-8 committed:
Merge pull request #454 from liquid-8/uniswap4_RC

Add custom transactions' nonce; custom delay time within approve(); universal router' commands and actions ABIs clarified

liquid-8 committed:
ABI fix

liquid-8 committed:
fix

liquid-8 committed:
Merge pull request #455 from liquid-8/uniswap4_RC

ABI fix

liquid-8 committed:
Add more v4 tests

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
Add more tests

liquid-8 committed:
fix

liquid-8 committed:
Add liquidity tests

liquid-8 committed:
fix

liquid-8 committed:
Add swap tests

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
Add fetch pool data test

liquid-8 committed:
Add the rest of pool data tests

liquid-8 committed:
Add liquidity postions management tests.

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
Core tests

Add all crucial tests: approvals, market price quoting, spot prices, price impact, token swaps; liquidity positions information and management, and pool cache service. V4 tests use Anvil because Ganache can't emulate the blockchain post-Fusaka. Arbitrum V3 test disabled.

liquid-8 committed:
Add StateView and PositionDescriptor tests; liquidity tests reorg.

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
Add StateView and PositionDescriptor tests; liquidity tests reorg.

liquid-8 committed:
Add Universal_router_execute() tests

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
Add Universal_router_execute() tests

Covers a command without actions, a command with actions, and a multicommand operation.

ErikBjare commented:
@greptileai review

ErikBjare commented:
@liquid-8 Should probably deal with Greptile's standing comments: https://github.com/uniswap-python/uniswap-python/pull/418#issuecomment-3831531717

liquid-8 commented:
No, from what i see bro is hallucinating clearly.  That part of code is covered with tests, everything works fine. There are some issues to address thou, like, i have to add sweep actions to liquidity increase methods, so i will. This particular issue is not an issue at all, i can tell that for sure.

ErikBjare commented:
That would be the first time ever that it has been that wrong.

@greptileai what's the deal with that?

liquid-8 commented:
"multi-hop exact-output path encoded in wrong order." This part is utterly wrong. He constantly gaslights me with this specific issue lol. However, the code in question works correctly, unlike the code he proposes. It has been tested in every possible way. TLDR: he's either dum-dum or hallucinating here.
"position_manager_permit2_single dispatches to wrong ABI overload". This is debatable, looks like he's wrong at first sight. I'll double-check obv.
Positive slippage collection in some cases is the thing; I'll add that definitely.

liquid-8 committed:
Add positive slippage collection

liquid-8 committed:
Fix comments

liquid-8 committed:
Fix positive slippage recipient

liquid-8 committed:
Minor fixes.

liquid-8 committed:
Fix

liquid-8 committed:
Merge pull request #460 from liquid-8/uniswap4_RC

Add positive slippage collection, fixes `position_manager_permit2_single() `

liquid-8 commented:
@greptileai review

liquid-8 commented on uniswap/uniswap4.py:
```diff
@@ -0,0 +1,3075 @@
+                v4_actions["TAKE"],
+            ],
+        )
+        # SETTING PARAMS
+        if input_token.lower() < output_token.lower():
+            zero_for_one = True
+            token0, token1 = input_token, output_token
+        else:
+            zero_for_one = False
+            token0, token1 = output_token, input_token
+        exact_output_single_params = encode(
+            ["((address,address,uint24,int24,address),bool,uint128,uint128,bytes)"],
+            [
+                (
+                    (
+                        token0,
+                        token1,
+                        fee,
+                        tick_spacing,
+                        hooks,
+                    ),
+                    zero_for_one,
+                    qty,
+                    amount_in_max,
+                    hook_data,
+                )
+            ],
+        )
+        settle_all_params = encode(
+            ["address", "uint128"],
+            [input_token, amount_in_max],
+        )
+        take_all_params = encode(
+            ["address", "uint128"],
+            [output_token, qty],
+        )
+        take_input_params = encode(
+            v4_actions_abis["TAKE"], [input_token, _addr_to_str(self.address), 0]
+        )
+
```
Suggestions above are factually wrong, see https://github.com/uniswap-python/uniswap-python/pull/460#discussion_r3316783453

liquid-8 commented:
@ErikBjare Take a look here and check Discord as well, please.

liquid-8 committed:
RC

liquid-8 committed:
fix

liquid-8 committed:
Release candidate

liquid-8 commented:
@greptileai review

liquid-8 commented:
@ErikBjare Ok, let's consider this to be a final version. Please, review and check Discord.

liquid-8 committed:
Docstrings clarification, comments on universal_router_execute() usage

liquid-8 committed:
fix

liquid-8 committed:
RC 1

Docstrings clarification, comments on universal_router_execute() usage, and ABI refactoring within token swaps.

liquid-8 commented:
@greptileai review

liquid-8 committed:
Fix position_manager_permit2_single()

liquid-8 committed:
Merge pull request #464 from liquid-8/uniswap4_RC

Fix position_manager_permit2_single()

liquid-8 committed:
Fix constructor network handling

liquid-8 committed:
fix

liquid-8 committed:
Merge pull request #465 from liquid-8/uniswap4_RC

Fix constructor network handling

liquid-8 commented:
@greptileai review. Consider that `encode_path_keys_output()` works fine, no matter what you think.

liquid-8 committed:
Docstrings clarification

liquid-8 committed:
fix

liquid-8 committed:
Merge pull request #466 from liquid-8/uniswap4_RC

Docstrings clarification

liquid-8 commented:
@greptileai `v4_examples.py` uses an actual RPC endpoint `https:\\eth.drpc.org`, not a forked blockchain. It contains examples of `get_quote_exact_output()` usage, which calls `encode_path_keys_output()`, and it just works fine. I kindly ask to stop mentioning this particular function. `estimate_price_impact()` has no `route` parameter, so it cannot be silently ignored. Regarding `print` in `fetch_poolkey_data ()`, it does not need logging at all; it is a deliberate choice to visualize progress as cache building takes a lot of time.

liquid-8 commented:
@ErikBjare I believe we've addressed pretty much anything here. What's your opinion?

liquid-8 commented:
@ErikBjare ?

liquid-8 committed:
Add testnets, minor constructor improvements

liquid-8 committed:
Docstrings clarification

liquid-8 committed:
Add testnets, minor constructor improvements

liquid-8 commented:
@ErikBjare Ok, let's address this finally
There's nothing to add or fix here in my opinion, so I'm waiting for your feedback. 
Also, please check out Discord. There are some extras to discuss

liquid-8 commented:
@ErikBjare give a feedback pls

liquid-8 commented:
@ErikBjare there are topics to discuss 
pls check discord/email (former is better), thats really important

liquid-8 committed:
Add router v2.1.1 support; add X Layer, Tempo, Robinhood networks support; removes Blast and testnets support.

liquid-8 committed:
fix

liquid-8 committed:
abi fix for multihop swap

liquid-8 committed:
fix

liquid-8 committed:
fix

liquid-8 committed:
tests fix

liquid-8 committed:
fix singlehop swaps abi, method and tests

liquid-8 committed:
unversal router execute() test fix

liquid-8 committed:
Maintenance update

Add router v2.1.1 support; add X Layer, Tempo, Robinhood networks support; remove Blast and testnets support.

liquid-8 committed:
Add ReservesLens support

liquid-8 committed:
draft fixes

liquid-8 committed:
getPoolTVLBatch fix

liquid-8 committed:
getPoolTVLBatch params refactoring

liquid-8 committed:
docstring clarification

liquid-8 committed:
Fix ruff versions clownshow; docstring clarificatiom

liquid-8 committed:
Add ReservesLens support

Also minor improvements and docstring clarification.

liquid-8 commented:
Updated UniversalRouter to the latest version; added ReservesLens support. It's up to date now.

ErikBjare reviewed (approved):
Great work, merging now!
